### DISCUSSION TITLE
Having trouble building and uploading to device

### DISCUSSION THREAD

**Discussion by Boingboingsplat** [boozook/playdate#230] [Q&A]
This project is looking really promising and I'm very interested! I'm having trouble with building and uploading examples to my device. I'm running things from Arch Linux.

I can use the commands from the example readme without any issues, which starts up the simulator and allows me to test the example. Uploading game to device from simulator results in `Couldn't find pdz file main.pdz` on device, but I expected to need to build for device using cargo-playdate anyways.

However, when I add `--device` and build it, the build succeeds, but when it comes to running the build I get `error: Playdate device not found`. Is there some setup step to allow cargo-playdate to see my device that I'm missing? The simulator is able to access the device without issue.

Full command I'm trying is:
`cargo playdate run -p=playdate-controls --example=accelerometer --features=sys/lang-items,sys/entry-point --device`

**Comment by jcornaz** [comment]
Here's the [just](https://just.systems/) recipes I used for one of my project.

It's a bit dated, but hopefully, it still works, or at least can give you some pointers to find a solution. 


The `just build` recipe builds the release artifact for both simulator and device:
```just
build:
    # build for simulator
    cargo playdate package --release --locked
    
    # build for device
    cargo playdate package --release --device --locked
```

Then the `just package` recipe (which will first run the build recipe) package it into a zip file like this:

(replace `{{project_name}}` by the name of your project as defined in your `Cargo.toml`)

```just
package: build
    # Copy the PDX to the root directory
    cp -r target/thumbv7em-none-eabihf/release/playdate/{{project_name}}.pdx .
    
    # Copy the Linux library into the PDX, so that it contains both the `.bin` file for device and `.so` for simulator
    cp target/release/playdate/{{project_name}}.pdx/pdex.so {{project_name}}.pdx/
    
    # Strib debug symbol to make the Linux library smaller
    strip {{project_name}}.pdx/pdex.so
    
    # Add a build number to the pdxinfo, I decided to have the build number being the number of commit in the git history
    echo "buildNumber=$(git log --oneline | wc -l)" >> {{project_name}}.pdx/pdxinfo
    
    # Zip the result
    zip -r {{project_name}}.pdx.zip {{project_name}}.pdx
```

The resulting zip file works for both Linux simulator and device. The device will simply ignore the `.so` and the simulator will ignore the `.bin`. My setup doesn't work on non-linux simulators.

I haven't tried it for a while, but it was definitely working, and I am also on archlinux ;-)

I hope it helps

**Reply by Boingboingsplat** [reply]
That's definitely useful knowledge to know about making a package that works in both the simulator and device! I'll probably setup something similar (just looks pretty nice) for myself.

I would like to know if there's a reason why cargo-playdate can't find my device anyways, though.

**Reply by boozook** [reply]
> I would like to know if there's a reason why cargo-playdate can't find my device anyways, though.

Currently that part in refactoring and already have to be able to find your device, but it isn't included in `cargo-playdate` yet.
May I ask you to test it on your system, please?

Branch: [refactor/pdtool](https://github.com/boozook/playdate/tree/refactor/pdtool) if you could install from git with
```bash
cargo +nightly install --git=https://github.com/boozook/playdate.git --branch=refactor/pdtool playdate-tool2
```

[Artifacts](https://github.com/boozook/playdate/actions/runs/8426605562#artifacts) direct links:
- [pdtool-macOS-X64](https://github.com/boozook/playdate/actions/runs/8426605562/artifacts/1357142365)
- [pdtool-Linux-X64](https://github.com/boozook/playdate/actions/runs/8426605562/artifacts/1357140765)
- [pdtool-Windows-X64.exe](https://github.com/boozook/playdate/actions/runs/8426605562/artifacts/1357142655)

What to test:
```bash
# connect your playdate __and unlock__

# your device(s) should be listed in stdout
pdtool list

# mount your device(s) as drive
pdtool mount

# turn device(s) back to normal mode
pdtool unmount --wait

# try other options described in help - `pdtool [command] --help`
```

Also could be great to try `pdtool install ...` on Windows 🙄.

**Comment by boozook** [comment]
> Uploading game to device from simulator results in `Couldn't find pdz file main.pdz` on device, but I expected to need to build for device using cargo-playdate anyways.

Builds for sim and device are different, definitely. So there are two options:
1. Build again for `--device` target
2. Build "two target into one package" combining two targets - `...build --simulator --device...`


- - -

> However, when I add `--device` and build it, the build succeeds, but when it comes to running the build I get `error: Playdate device not found`. Is there some setup step to allow cargo-playdate to see my device that I'm missing? The simulator is able to access the device without issue.

What OS you're using?
Could you try `pdtool` from [my comment above](https://github.com/boozook/playdate/discussions/230#discussioncomment-8913114), please, and tell me is that able to find your device and install/run too?



- - -

Possible workaround for now:
May be help to build `cargo-playdate` __with__ or __without__ `usb` feature.

**Reply by Boingboingsplat** [reply]
Thanks for all the information! I hadn't realized that `--simulator` and `--device` could be combined as options, very handy!

I'm using Arch Linux as my OS.

I tested `pdtool` from your comment. I was able to mount it as a file system... though notably it took many tries. Unmounting failed with an error, and thus the `--wait` timed out.

```
$ pdtool mount
[2024-03-26T20:47:35Z DEBUG pdtool] cmd: Mount {
        query: DeviceQuery {
            value: None,
        },
        wait: false,
    }
[2024-03-26T20:47:35Z DEBUG playdate_device::mount::methods] mounting PDU1-Y042992
[2024-03-26T20:47:35Z TRACE playdate_device::mount::methods] create sending fut
[2024-03-26T20:47:35Z TRACE playdate_device::serial::discover] found port: /dev/ttyACM0, dev sn: Some("PDU1-Y042992")
[2024-03-26T20:47:35Z TRACE playdate_device::serial::discover] sn is ≈ true
[2024-03-26T20:47:35Z TRACE playdate_device::serial::r#async] sending `datadisk` to /dev/ttyACM0
[2024-03-26T20:47:35Z DEBUG playdate_device::usb] closing PDU1-Y042992

[ Duplicate tries clipped... ]

[2024-03-26T20:47:35Z TRACE playdate_device::device::methods] try: 14/130
[2024-03-26T20:47:35Z TRACE playdate_device::device::methods] waiting mode S of PDU1-Y042992, current: D
[2024-03-26T20:47:35Z DEBUG playdate_device::usb] PDU1-Y042992: refreshed by existing new.

[ Duplicate tries clipped... ]

[2024-03-26T20:47:38Z TRACE playdate_device::mount::methods] try: 30/100
[2024-03-26T20:47:38Z TRACE playdate_device::mount::methods] waiting mount point availability: PDU1-Y042992, current: S
[2024-03-26T20:47:38Z DEBUG playdate_device::mount::methods] PDU1-Y042992 mounted, volume found: '/run/media/boing/PLAYDATE'
PDU1-Y042992 (S) /run/media/boing/PLAYDATE
```

Unmounted resulted in some errors followed by a timeout. I had to manually reset my device to get it out of its data disk mode.

```
$ pdtool unmount --wait
[2024-03-26T20:47:50Z DEBUG pdtool] cmd: Unmount {
        query: DeviceQuery {
            value: None,
        },
        wait: true,
    }
[2024-03-26T20:47:50Z TRACE playdate_device::mount::methods] Unmounting PDU1-Y042992
[2024-03-26T20:47:50Z TRACE playdate_device::mount::methods] Unmounting PDU1-Y042992: /run/media/boing/PLAYDATE
[2024-03-26T20:47:50Z TRACE playdate_device::mount::methods] Unmounting 1 devices.
[2024-03-26T20:47:50Z DEBUG playdate_device::usb] closing PDU1-Y042992
eject: cannot open /dev/sdb: Permission denied
umount: /run/media/boing/PLAYDATE: No such file or directory
Error unmounting /dev/sdb1: GDBus.Error:org.freedesktop.UDisks2.Error.NotMounted: Device `/dev/sdb1' is not mounted
[ Duplicate tries for --wait clipped, as the device never changed mode... ]
```

Notably, running `sudo pdtool unmount` allows it to perform as expected.

**Reply by boozook** [reply]
> Thanks for all the information! 
> ...very handy!

Thank you! 😊 

> I'm using Arch Linux as my OS.

> I was able to mount it as a file system... though notably it took many tries. 

Sorry for mess of logs. That's "heavy debug mode".


> Unmounting failed with an error, and thus the `--wait` timed out.

Thanks! 🙏 
Timeouts will be configurable;

> Unmounted resulted in some errors followed by a timeout. I had to manually reset my device to get it out of its data disk mode.

Ok, I know the problem. The program or user haven't permitted to eject (with usb power reset) usb.

I have to write pre-check for this case or find some workaround that doesn't require sudo. Do you have any suggestions?

> Notably, running `sudo pdtool unmount` allows it to perform as expected.

Yup. I maybe should pre-check permissions and print a hint for user to add self to telmodem (or whatever) group.
And anyway also hint about "Press A on the device to reboot in normal mode".

**Comment by boozook** [comment]
[Here](https://github.com/boozook/playdate/actions/runs/8492822926) (also #231) is new build of pdtool. @jcornaz, @Boingboingsplat, please could you try it on your system with connected device? It will be incredibly helpful. 🥹

**Reply by boozook** [reply]
I hope many issues has been solved. Could you guys try new versions, plz?
[There is latest prerelease with attached binaries](https://github.com/boozook/playdate/releases/tag/2024.04.02).
@jcornaz 
@Boingboingsplat

**Comment by boozook** [comment]
@Boingboingsplat, would you please update to latest versions and give some feedback - is problem solved for you?
